static/src/javascripts/projects/commercial/modules/messenger/scroll.js

```diff
@@ -0,0 +1,125 @@
+define([
+    'common/utils/closest',
+    'common/utils/fastdom-promise',
+    'commercial/modules/messenger'
+], function (closest, fastdom, messenger) {
+    // An intersection observer will allow us to efficiently send slot
+    // coordinates for only those that are in the viewport.
+    var useIO = 'IntersectionObserver' in window;
+    var taskQueued = false;
+    var listeners = {};
+    var slots = {};
+    var listenerCounter = 0;
+    var observer;
+
+    messenger.register('scroll', onMessage, { persist: true });
+
+    function onMessage(respond, start, iframe) {
+        if (start) {
+            addScrollListener(iframe.id, closest(iframe, '.js-ad-slot'), respond);
+        } else {
+            removeScrollListener(iframe.id);
+        }
+    }
+
+    return {
+        addScrollListener: addScrollListener,
+        removeScrollListener: removeScrollListener
+    };
+
+    function addScrollListener(id, slot, respond) {
+        if (listenerCounter === 0) {
+            window.addEventListener('scroll', onScroll);
+            if (useIO) {
+                observer = new window.IntersectionObserver(onIntersect);
+            }
+        }
+
+        listeners[id] = {
+            slot: slot,
+            // When using IOs, a slot is hidden by default. When the IO starts
+            // observing it, the onIntercept callback will be triggered if it
+            // is already in the viewport
+            visible: !useIO,
+            respond: respond
+        };
+        listenerCounter += 1;
+
+        if (useIO) {
+            slots[slot.id] = {
+                slot: slot,
+                iframeId: id
+            };
+            observer.observe(slot);
+        }
+    }
+
+    function removeScrollListener(id) {
+        if (listeners[id]) {
+            if (useIO) {
+                observer.unobserve(listeners[id].slot);
+                slots[listeners[id].slot.id] = false;
+            }
+            listeners[id] = false;
+            listenerCounter -= 1;
+        }
+
+        if (listenerCounter === 0) {
+            window.removeEventListener('scroll', onScroll);
+            if (useIO) {
+                observer.disconnect();
+                observer = null;
+            }
+        }
+    }
+
+    function onScroll() {
+        if (!taskQueued) {
+            taskQueued = true;
+            fastdom.read(function () {
+                taskQueued = false;
+
+                var slots = Object.keys(listeners)
+                    .filter(function (id) {
+                        return listeners[id].visible;
```

Comment:
Yes, that makes sense, i suspected it was because `visible` was set to `!useIO`. I thought you could rejig the code so that there is a single `filter` call, and inside the filter function, if `useIO` is `false`, you use a viewport-top check, and if `true`, it uses `visible`. The source of the filter would be `Object.keys(listeners)`, rather than [slots](https://github.com/guardian/frontend/pull/14155/files/0f2bc13a4cfd5e2e2375d2e7527f1e7c2f470328#diff-2fb78cc4b5143edecc6410391f4cbd3eR93)